### ISSUE TITLE
Running Dynamo UI-less appears to have thread bugs

### ISSUE THREAD

**Issue by ThomasMahon** [DynamoDS/DynamoRevit#2639]
If this issue is **not** a bug report or improvement request, please check the [Dynamo forum](https://forum.dynamobim.com/), and start a thread there to discuss your issue.

## Dynamo version

Dynamo 2.02

## Revit version

R2019

## Operating system

Win10

## What did you do? 

Created a Revit-addin which executes Dynamo scripts via Revit buttons. We're using the DynamoRevitDS library and initializing a `DynamoRevit` object to pass in the journal file to create the `DynamoRevitCommandData` which is executed when the Revit button is clicked. 

Some of the scripts have UI nodes from packages such as DataShapes and this is how we know the scripts are executing erroneously since the UIs display when the ribbon button is clicked, and unexpectedly display again typically 2 times, if the user does anything else after clicking the button to run the script (such as clicking a different button, or modify the model). We dismiss these pop-ups and that seems to clear the problem, i.e. they do not show again, until we click the button to run the script again and this unexpected behaviour occurs again.

Its almost like there is some asynconous event Dynamo is performing and its being raised more than once when it should only be raised once per button click/dynamo script execution. If we try to debug for example, the unexpected executions of the graph never hit any break points in our app; this is what is leading us to believe its a problem with Dynamo running in UI-less mode.

We've done a lot of testing to try and rule out any errors in our application or in the graphs which may be the culprit but there is nothing which appears to be triggering the multiple executions. Some more detail:

1. When  JournalKeys.AutomationModeKey  = false    and    when JournalKeys.ForceManualRunKey = true
- .dyn files which have any nodes that show RevitAPIUI TaskDialog to the user, execute normally. By normally I mean that the messages don't fire again after the .dyn file is ran and any changes are made to the Revit document afterwords by the user.
- .dyn files which use Winforms, such as Data Shapes which ask for user input before running the script, do not execute normally. For example, after the script is executed successfully, and returns control to the caller, then Data Shape nodes take action and show their UI making them totally out of sync, because the line of code that opens up dynamo and executes the script has already been called by the executing thread.

2. When  JournalKeys.AutomationModeKey  = true    and    when JournalKeys.ForceManualRunKey = false
 - Any script containing either a node which shows a RevitAPIUI TaskDialog or a Winform, do not execute normally. By normally I mean, that after the script is executed successfully and the  buttons execute method is popped of the stack, both TaskDialog's and  Winforms instantiated by custom nodes keep showing their UI after any changes are made to the current Revit document by the user.

Is this a bug or limitation in running Dynamo in this way? What is also strange is the Dynamo Player does not have this problem running the same scripts. What is different between how Dynamo Player executes scripts and how using the DynamoRevit object executes scripts?

@mjkkirschner @Amoursol @aparajit-pratap your input would be greatly appreciated as finding a solution for this is business critical. 

## What did you expect to see?

The script executed once.

## What did you see instead?

See above.

**Comment by mjkkirschner** [comment]
Hi @ThomasMahon - before diving too far, what you describe seems as designed. If you run a graph in automatic mode, you're starting up a DynamoModel, then just like if the UI was present, attaching a bunch of handlers and updaters that will continue to execute the graph when the document changes.

It's possible some logic in the UI layer is now missing which might have prevented double executions in some cases as well.

What I think you should try, is to shutdown the DynamoModel after the graph is executed.

**Comment by mjkkirschner** [comment]
The other issue you should consider, is that DynamoRevit and DynamoModel are **NOT** currently designed in a way that supports multiple instances of either the same process, appdomain etc.

You should be able to reuse the existing DynamoModel which is loaded by Revit.

If you cannot or you must instantiate your own for some reason you must isolate it from the existing one.

**Comment by ThomasMahon** [comment]
Hi @mjkkirschner 

Thanks for the response, its very helpful. I didn't find a reliable way to obtain an instance of the DynamoModel and shutting down the RevitDynamoModel cuased exceptions and prevented Dynamo from running again in the same Revit session. 

Instead, we just set all the graphs to manual and call ForceRun() on the RevitDynamoModel and that seems to do what we need. 

Any more information or documentation you can provide for running Dynamo as a background service would be greatly appreciated as we are currently having to resort to a lot of trial and error and hit many deadends.

**Comment by aparajit-pratap** [comment]
@sm6srw any ideas on what could be different in the execution model of Player vs. Dynamo that could cause multiple executions in Dynamo but not in Player?

**Comment by johnpierson** [comment]
Stumbled upon this thread as I just ran into this with a user of Data Shapes nodes in UI-less dynamo.

This _seemed_ to fix it for me: https://github.com/ParallaxTeam/Relay/blob/05b815f1519b6c2a942db2960945108bc187959e/src/Revit/Command.cs#L134

**Comment by ThomasMahon** [comment]
Yup, I know! The RevitDynamoModel is a singleton so if you dispose it you basically dispose the Dynamo application (and with it the ability to run any scripts and that includes the ones set to ribbon buttons) and there's no way to reinitialize it without restarting Revit - this also explains why Dynamo can only support one tab/file being open at any 1 time incidentally. The ForceRun() technique is the best option for now.
